src/dictionary.rs

```diff
@@ -0,0 +1,445 @@
+/// ]);
+///
+/// for phrase in dict.lookup_word(
+///     syl![Bopomofo::C, Bopomofo::E, Bopomofo::TONE4]
+/// ) {
+///     assert_eq!("測", phrase.as_str());
+///     assert_eq!(100, phrase.freq());
+/// }
+/// ```
+pub type Phrases<'a, 'p> = Box<dyn Iterator<Item = Phrase<'p>> + 'a>;
+
+pub type DictEntries<'a, 'p> = Box<dyn Iterator<Item = (Vec<Syllable>, Phrase<'p>)> + 'a>;
+
+/// An interface for looking up dictionaries.
+///
+/// This is the main dictionary trait. For more about the concept of
+/// dictionaries generally, please see the [module-level
+/// documentation][crate::dictionary].
+///
+/// # Examples
+///
+/// The std [`HashMap`] implements the `Dictionary` trait so it can be used in
+/// tests.
+///
+/// ```
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// use std::collections::HashMap;
+///
+/// use chewing::{dictionary::Dictionary, syl, zhuyin::Bopomofo};
+///
+/// let mut dict = HashMap::new();
+/// let dict_mut = dict.as_mut_dict().unwrap();
+/// dict_mut.insert(&[syl![Bopomofo::C, Bopomofo::E, Bopomofo::TONE4]], ("測", 100).into())?;
+///
+/// for phrase in dict.lookup_word(
+///     syl![Bopomofo::C, Bopomofo::E, Bopomofo::TONE4]
+/// ) {
+///     assert_eq!("測", phrase.as_str());
+///     assert_eq!(100, phrase.freq());
+/// }
+/// # Ok(())
+/// # }
+/// ```
+pub trait Dictionary: Debug {
+    /// Returns an iterator to all single syllable words matched by the
+    /// syllable, if any. The result should use a stable order each time for the
+    /// same input.
+    fn lookup_word(&self, syllable: Syllable) -> Phrases {
+        self.lookup_phrase(&[syllable])
+    }
+    /// Returns an iterator to all phrases matched by the syllables, if any. The
+    /// result should use a stable order each time for the same input.
+    fn lookup_phrase(&self, syllables: &[Syllable]) -> Phrases;
+    /// Returns an iterator to all phrases in the dictionary.
+    fn entries(&self) -> DictEntries;
+    /// Returns information about the dictionary instance.
+    fn about(&self) -> DictionaryInfo;
+    /// Returns a mutable reference to the dictionary if the underlying
+    /// implementation allows update.
+    fn as_mut_dict(&mut self) -> Option<&mut dyn DictionaryMut>;
+}
+
+/// An interface for updating dictionaries.
+///
+/// For more about the concept of dictionaries generally, please see the
+/// [module-level documentation][crate::dictionary].
+///
+/// # Examples
+///
+/// The std [`HashMap`] implements the `DictionaryMut` trait so it can be used in
+/// tests.
+///
+/// ```
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// use std::collections::HashMap;
+///
+/// use chewing::{dictionary::Dictionary, syl, zhuyin::Bopomofo};
+///
+/// let mut dict = HashMap::new();
+/// let dict_mut = dict.as_mut_dict().unwrap();
+/// dict_mut.insert(&[syl![Bopomofo::C, Bopomofo::E, Bopomofo::TONE4]], ("測", 100).into())?;
+/// # Ok(())
+/// # }
+/// ```
+pub trait DictionaryMut {
+    fn insert(
```

Comment:
I hope there are documents for these trait methods (insert/update/remove). Not that difficult for me to look into implementation myself. But it would be nice to know, for example, `remove` will remove one of syllables' phrases if possible at first glance.

Reply:
Agree. Let's review all doc comments after the interfaces are somewhat stable.